### PR TITLE
Custom binaries for dynamic analysis

### PR DESCRIPTION
**Pull request by jssblck**
# Overview

Adds the concept of overriding the binary used for analysis to the CLI, first used for maven here.
In the future we can add more binaries as desired.

Also adds `mvnw` as an option when analyzing maven projects.

## Acceptance criteria

- We are able to analyze maven projects as before.
- We are able to analyze maven projects that require `mvnw`.
- We are able to provide a custom maven binary to use.

## Testing plan

Relied on automated testing, plus:

- Validated that custom command fallback works by providing `FOSSA_MAVEN_CMD=go`:

<details>
<summary>Using 'go' as an override command</summary>

```
; wget https://github.com/google/guava/archive/refs/tags/v31.1.tar.gz
; tar -xzf v31.1.tar.gz
; FOSSA_MAVEN_CMD=go cabal run fossa -- analyze ~/projects/scratch/ -o
Up to date
[ INFO] Analyzing maven project at /home/jess/projects/scratch/guava-31.1/android/guava-bom/
[ INFO] Analyzing maven project at /home/jess/projects/scratch/guava-31.1/futures/failureaccess/
[ INFO] Analyzing maven project at /home/jess/projects/scratch/guava-31.1/
[ INFO] Analyzing maven project at /home/jess/projects/scratch/guava-31.1/guava-bom/
[ INFO] Analyzing maven project at /home/jess/projects/scratch/guava-31.1/futures/listenablefuture9999/
[ INFO] Analyzing maven project at /home/jess/projects/scratch/guava-31.1/futures/listenablefuture1/
[ INFO] Analyzing maven project at /home/jess/projects/scratch/guava-31.1/android/
[ INFO]
[ INFO] Scan Summary
[ INFO] ------------
[ INFO] fossa-cli branch custom-binaries-dynamic-analysis (revision 2b048f43d1b7 compiled with ghc-9.0)
[ INFO] fossa endpoint server version: N/A
[ INFO]
[ INFO] 7 projects scanned;  0 skipped,  7 succeeded,  0 failed,  42 analysis warnings
[ INFO]
[ INFO] * maven project in "/home/jess/projects/scratch/guava-31.1/": succeeded with 6 warnings
[ INFO] * maven project in "/home/jess/projects/scratch/guava-31.1/android/": succeeded with 6 warnings
[ INFO] * maven project in "/home/jess/projects/scratch/guava-31.1/android/guava-bom/": succeeded with 6 warnings
[ INFO] * maven project in "/home/jess/projects/scratch/guava-31.1/futures/failureaccess/": succeeded with 6 warnings
[ INFO] * maven project in "/home/jess/projects/scratch/guava-31.1/futures/listenablefuture1/": succeeded with 6 warnings
[ INFO] * maven project in "/home/jess/projects/scratch/guava-31.1/futures/listenablefuture9999/": succeeded with 6 warnings
[ INFO] * maven project in "/home/jess/projects/scratch/guava-31.1/guava-bom/": succeeded with 6 warnings
[ INFO] -
[ INFO]
[ INFO]   Some projects may not appear in the summary if they were filtered during discovery.
[ INFO]   You can run `fossa list-targets` to see all discoverable projects.
[ INFO]
[ INFO] You can pass `--debug` option to eagerly show all warning and failure messages.
[ INFO] You can also view analysis summary with warning and error messages at: "/tmp/fossa-analyze-scan-summary.txt"
[ INFO] ------------
{ ... output snipped ... }

; cat /tmp/fossa-analyze-scan-summary.txt
Scan Summary
------------
fossa-cli branch custom-binaries-dynamic-analysis (revision 2b048f43d1b7 compiled with ghc-9.0)
fossa endpoint server version: N/A

7 projects scanned;  0 skipped,  7 succeeded,  0 failed,  42 analysis warnings

* maven project in "/home/jess/projects/scratch/guava-31.1/": succeeded with 6 warnings
* maven project in "/home/jess/projects/scratch/guava-31.1/android/": succeeded with 6 warnings
* maven project in "/home/jess/projects/scratch/guava-31.1/android/guava-bom/": succeeded with 6 warnings
* maven project in "/home/jess/projects/scratch/guava-31.1/futures/failureaccess/": succeeded with 6 warnings
* maven project in "/home/jess/projects/scratch/guava-31.1/futures/listenablefuture1/": succeeded with 6 warnings
* maven project in "/home/jess/projects/scratch/guava-31.1/futures/listenablefuture9999/": succeeded with 6 warnings
* maven project in "/home/jess/projects/scratch/guava-31.1/guava-bom/": succeeded with 6 warnings
-
----------
maven project in "/home/jess/projects/scratch/guava-31.1/": succeeded with 6 warnings

Warning

  Command "mvnw" not suitable: running with args ["-v"] resulted in a non-zero exit code

  >>> Relevant errors

    Error

      Could not find executable: `mvnw`.

      Please ensure `mvnw` exist in PATH prior to running fossa.

      If you believe this to be a defect, please report a bug to FOSSA support at https://support.fossa.com

      Traceback:
        - Running command 'mvnw'
        - Evaluate command: mvnw
        - Evaluate command: go
        - Command override provided: go
        - Make analysis command from CandidateAnalysisCommands {candidateCmdNames = "mvnw" :| ["mvn"], candidateCmdArgs = ["-v"], candidateOverrideKind = Just MavenType}
        - Running plugin to get dependency graph
        - Plugin analysis
        - Dynamic Analysis
        - Maven
        - Project Analysis: MavenProjectType



Warning

  Command "go" not suitable: running with args ["-v"] resulted in a non-zero exit code

  >>> Relevant errors

    Error

      Command execution failed:
          command: Command {cmdName = "go", cmdArgs = ["-v"], cmdAllowErr = Never}
          dir: /home/jess/projects/scratch/guava-31.1/
          exit: ExitFailure 2
          stdout:

          stderr:
            flag provided but not defined: -v
            Go is a tool for managing Go source code.

            Usage:

            	go <command> [arguments]

            The commands are:

            	bug         start a bug report
            	build       compile packages and dependencies
            	clean       remove object files and cached files
            	doc         show documentation for package or symbol
            	env         print Go environment information
            	fix         update packages to use new APIs
            	fmt         gofmt (reformat) package sources
            	generate    generate Go files by processing source
            	get         add dependencies to current module and install them
            	install     compile and install packages and dependencies
            	list        list packages or modules
            	mod         module maintenance
            	work        workspace maintenance
            	run         compile and run Go program
            	test        test packages
            	tool        run specified go tool
            	version     print Go version
            	vet         report likely mistakes in packages

            Use "go help <command>" for more information about a command.

            Additional help topics:

            	buildconstraint build constraints
            	buildmode       build modes
            	c               calling between Go and C
            	cache           build and test caching
            	environment     environment variables
            	filetype        file types
            	go.mod          the go.mod file
            	gopath          GOPATH environment variable
            	gopath-get      legacy GOPATH go get
            	goproxy         module proxy protocol
            	importpath      import path syntax
            	modules         modules, module versions, and more
            	module-get      module-aware go get
            	module-auth     module authentication using go.sum
            	packages        package lists and patterns
            	private         configuration for downloading non-public code
            	testflag        testing flags
            	testfunc        testing functions
            	vcs             controlling version control with GOVCS

            Use "go help <topic>" for more information about that topic.


      If you believe this to be a defect, please report a bug to FOSSA support at https://support.fossa.com

      Traceback:
        - Running command 'go'
        - Evaluate command: go
        - Command override provided: go
        - Make analysis command from CandidateAnalysisCommands {candidateCmdNames = "mvnw" :| ["mvn"], candidateCmdArgs = ["-v"], candidateOverrideKind = Just MavenType}
        - Running plugin to get dependency graph
        - Plugin analysis
        - Dynamic Analysis
        - Maven
        - Project Analysis: MavenProjectType



Warning

  Command "mvnw" not suitable: running with args ["-v"] resulted in a non-zero exit code

  >>> Relevant errors

    Error

      Could not find executable: `mvnw`.

      Please ensure `mvnw` exist in PATH prior to running fossa.

      If you believe this to be a defect, please report a bug to FOSSA support at https://support.fossa.com

      Traceback:
        - Running command 'mvnw'
        - Evaluate command: mvnw
        - Evaluate command: go
        - Command override provided: go
        - Make analysis command from CandidateAnalysisCommands {candidateCmdNames = "mvnw" :| ["mvn"], candidateCmdArgs = ["-v"], candidateOverrideKind = Just MavenType}
        - Running plugin to get submodule names
        - Plugin analysis
        - Dynamic Analysis
        - Maven
        - Project Analysis: MavenProjectType



Warning

  Command "go" not suitable: running with args ["-v"] resulted in a non-zero exit code

  >>> Relevant errors

    Error

      Command execution failed:
          command: Command {cmdName = "go", cmdArgs = ["-v"], cmdAllowErr = Never}
          dir: /home/jess/projects/scratch/guava-31.1/
          exit: ExitFailure 2
          stdout:

          stderr:
            flag provided but not defined: -v
            Go is a tool for managing Go source code.

            Usage:

            	go <command> [arguments]

            The commands are:

            	bug         start a bug report
            	build       compile packages and dependencies
            	clean       remove object files and cached files
            	doc         show documentation for package or symbol
            	env         print Go environment information
            	fix         update packages to use new APIs
            	fmt         gofmt (reformat) package sources
            	generate    generate Go files by processing source
            	get         add dependencies to current module and install them
            	install     compile and install packages and dependencies
            	list        list packages or modules
            	mod         module maintenance
            	work        workspace maintenance
            	run         compile and run Go program
            	test        test packages
            	tool        run specified go tool
            	version     print Go version
            	vet         report likely mistakes in packages

            Use "go help <command>" for more information about a command.

            Additional help topics:

            	buildconstraint build constraints
            	buildmode       build modes
            	c               calling between Go and C
            	cache           build and test caching
            	environment     environment variables
            	filetype        file types
            	go.mod          the go.mod file
            	gopath          GOPATH environment variable
            	gopath-get      legacy GOPATH go get
            	goproxy         module proxy protocol
            	importpath      import path syntax
            	modules         modules, module versions, and more
            	module-get      module-aware go get
            	module-auth     module authentication using go.sum
            	packages        package lists and patterns
            	private         configuration for downloading non-public code
            	testflag        testing flags
            	testfunc        testing functions
            	vcs             controlling version control with GOVCS

            Use "go help <topic>" for more information about that topic.


      If you believe this to be a defect, please report a bug to FOSSA support at https://support.fossa.com

      Traceback:
        - Running command 'go'
        - Evaluate command: go
        - Command override provided: go
        - Make analysis command from CandidateAnalysisCommands {candidateCmdNames = "mvnw" :| ["mvn"], candidateCmdArgs = ["-v"], candidateOverrideKind = Just MavenType}
        - Running plugin to get submodule names
        - Plugin analysis
        - Dynamic Analysis
        - Maven
        - Project Analysis: MavenProjectType



Warning

  Command "mvnw" not suitable: running with args ["-v"] resulted in a non-zero exit code

  >>> Relevant errors

    Error

      Could not find executable: `mvnw`.

      Please ensure `mvnw` exist in PATH prior to running fossa.

      If you believe this to be a defect, please report a bug to FOSSA support at https://support.fossa.com

      Traceback:
        - Running command 'mvnw'
        - Evaluate command: mvnw
        - Evaluate command: go
        - Command override provided: go
        - Make analysis command from CandidateAnalysisCommands {candidateCmdNames = "mvnw" :| ["mvn"], candidateCmdArgs = ["-v"], candidateOverrideKind = Just MavenType}
        - Installing plugin
        - Plugin analysis
        - Dynamic Analysis
        - Maven
        - Project Analysis: MavenProjectType



Warning

  Command "go" not suitable: running with args ["-v"] resulted in a non-zero exit code

  >>> Relevant errors

    Error

      Command execution failed:
          command: Command {cmdName = "go", cmdArgs = ["-v"], cmdAllowErr = Never}
          dir: /home/jess/projects/scratch/guava-31.1/
          exit: ExitFailure 2
          stdout:

          stderr:
            flag provided but not defined: -v
            Go is a tool for managing Go source code.

            Usage:

            	go <command> [arguments]

            The commands are:

            	bug         start a bug report
            	build       compile packages and dependencies
            	clean       remove object files and cached files
            	doc         show documentation for package or symbol
            	env         print Go environment information
            	fix         update packages to use new APIs
            	fmt         gofmt (reformat) package sources
            	generate    generate Go files by processing source
            	get         add dependencies to current module and install them
            	install     compile and install packages and dependencies
            	list        list packages or modules
            	mod         module maintenance
            	work        workspace maintenance
            	run         compile and run Go program
            	test        test packages
            	tool        run specified go tool
            	version     print Go version
            	vet         report likely mistakes in packages

            Use "go help <command>" for more information about a command.

            Additional help topics:

            	buildconstraint build constraints
            	buildmode       build modes
            	c               calling between Go and C
            	cache           build and test caching
            	environment     environment variables
            	filetype        file types
            	go.mod          the go.mod file
            	gopath          GOPATH environment variable
            	gopath-get      legacy GOPATH go get
            	goproxy         module proxy protocol
            	importpath      import path syntax
            	modules         modules, module versions, and more
            	module-get      module-aware go get
            	module-auth     module authentication using go.sum
            	packages        package lists and patterns
            	private         configuration for downloading non-public code
            	testflag        testing flags
            	testfunc        testing functions
            	vcs             controlling version control with GOVCS

            Use "go help <topic>" for more information about that topic.


      If you believe this to be a defect, please report a bug to FOSSA support at https://support.fossa.com

      Traceback:
        - Running command 'go'
        - Evaluate command: go
        - Command override provided: go
        - Make analysis command from CandidateAnalysisCommands {candidateCmdNames = "mvnw" :| ["mvn"], candidateCmdArgs = ["-v"], candidateOverrideKind = Just MavenType}
        - Installing plugin
        - Plugin analysis
        - Dynamic Analysis
        - Maven
        - Project Analysis: MavenProjectType

# ... more of the same warnings for other targets in this directory omitted
```
</details>

- Validated that custom commands are preferred by providing `which mvn` as a custom command:

```
; FOSSA_MAVEN_CMD=`which mvn` cabal run fossa -- analyze ~/projects/scratch/ -o
Warning: The package list for 'hackage.haskell.org' is 27 days old.
Run 'cabal update' to get the latest list of available packages.
Resolving dependencies...
Up to date
[ INFO] Analyzing maven project at /home/jess/projects/scratch/guava-31.1/android/guava-bom/
[ INFO] Analyzing maven project at /home/jess/projects/scratch/guava-31.1/
[ INFO] Analyzing maven project at /home/jess/projects/scratch/guava-31.1/guava-bom/
[ INFO] Analyzing maven project at /home/jess/projects/scratch/guava-31.1/futures/listenablefuture9999/
[ INFO] Analyzing maven project at /home/jess/projects/scratch/guava-31.1/futures/listenablefuture1/
[ INFO] Analyzing maven project at /home/jess/projects/scratch/guava-31.1/futures/failureaccess/
[ INFO] Analyzing maven project at /home/jess/projects/scratch/guava-31.1/android/
[ INFO]
[ INFO] Scan Summary
[ INFO] ------------
[ INFO] fossa-cli branch custom-binaries-dynamic-analysis (revision 2b048f43d1b7 compiled with ghc-9.0)
[ INFO] fossa endpoint server version: N/A
[ INFO]
[ INFO] 7 projects scanned;  0 skipped,  7 succeeded,  0 failed,  0 analysis warnings
[ INFO]
[ INFO] * maven project in "/home/jess/projects/scratch/guava-31.1/": succeeded
[ INFO] * maven project in "/home/jess/projects/scratch/guava-31.1/android/": succeeded
[ INFO] * maven project in "/home/jess/projects/scratch/guava-31.1/android/guava-bom/": succeeded
[ INFO] * maven project in "/home/jess/projects/scratch/guava-31.1/futures/failureaccess/": succeeded
[ INFO] * maven project in "/home/jess/projects/scratch/guava-31.1/futures/listenablefuture1/": succeeded
[ INFO] * maven project in "/home/jess/projects/scratch/guava-31.1/futures/listenablefuture9999/": succeeded
[ INFO] * maven project in "/home/jess/projects/scratch/guava-31.1/guava-bom/": succeeded
[ INFO] -
[ INFO]
[ INFO]   Some projects may not appear in the summary if they were filtered during discovery.
[ INFO]   You can run `fossa list-targets` to see all discoverable projects.
[ INFO]
[ INFO] You can pass `--debug` option to eagerly show all warning and failure messages.
[ INFO] You can also view analysis summary with warning and error messages at: "/tmp/fossa-analyze-scan-summary.txt"
[ INFO] ------------
```

## Risks

The primary risk is that this implementation is confusing to users and opens a new can of worms for CLI configuration.

Secondarily, if this feature is buggy, users won't be happy.

## References

[ANE-778](https://fossa.atlassian.net/browse/ANE-778)

## Checklist

- [x] I added tests for this PR's change (or explained in the PR description why tests don't make sense).
- [ ] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [ ] If this change is externally visible, I updated `Changelog.md`. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [x] If I made changes to `.fossa.yml` or `fossa-deps.{json.yml}`, I updated `docs/references/files/*.schema.json`. You may also need to update these if you have added/removed new dependency type (e.g. `pip`) or analysis target type (e.g. `poetry`).

[ANE-778]: https://fossa.atlassian.net/browse/ANE-778?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/references/strategies/languages/maven/maven.md [hunk 1]
```diff
@@ -14,7 +14,15 @@ Maven analysis attempts these analysis methods in order:
 3. Run the maven plugin command version 3.3.0.
 4. Scan `pom.xml` files located in the file tree.
 
+FOSSA CLI uses [strategy command selection](../../../../features/strategy-command-selection.md) for Maven commands:
+- Maven _candidate commands_ are:
+  - If present, the command specified by the `FOSSA_MAVEN_CMD` environment variable.
+  - If present, the `mvnw` or `mvnw.bat` script in the project directory or any ancestor directory.
+  - Finally, the `mvn` command, which is searched for in `$PATH`.
+- The Maven _evaluation flag_ is `-v` for each _candidate command_.
```
Comment (zlav) [inline docs/references/strategies/languages/maven/maven.md:22]: I wouldn't change anything here, but I would add something like the following as a TLDR so user's don't have to read through the `strategy command selection` doc to understand what a `evaluation flag` is.
```
The FOSSA CLI tries `./mvnw -v` before moving onto `mvn -v` and selects the first one that succeeds to run the analysis tactic.
```

**Review by zlav — APPROVED** [review]
Docs/comments/tests look amazing 🔥 
Left one comment about how the analysis effects are being used but thats nonblocking. Looks good to merge after the comment about executing multiple commands is addressed.

File: src/Strategy/Maven/Plugin.hs [hunk 2]
```diff
@@ -191,58 +205,67 @@ textArtifactToPluginOutput
...
+-- Otherwise just use 'mvn'.
+mavenCmdCandidates :: (Has ReadFS sig m, Has Diagnostics sig m) => Path Abs Dir -> m CandidateAnalysisCommands
+mavenCmdCandidates dir = do
+  cmds <- (flip (:|) ["mvn"] . toText <$> mvnWrapperPath) <||> pure (NE.singleton "mvn")
+  pure $ CandidateAnalysisCommands cmds ["-v"] $ Just MavenType
+  where
+    -- Unlike with the gradle wrapper, it's not _expected_ for maven projects to use the maven wrapper.
+    -- Given that, don't warn on failure to find a wrapper; only warn if we find a wrapper and it fails to execute.
+    mvnWrapperPath :: (Has ReadFS sig m, Has Diagnostics sig m) => m (Path Abs File)
+    mvnWrapperPath = findFileInAncestor dir "mvnw" <||> findFileInAncestor dir "mvnw.bat"
```
[the remark this answers is not in the document]
Comment (jssblck) [reply src/Strategy/Maven/Plugin.hs:218]: Good point, will make that happen

**Comment by jssblck** [comment]
@zlav I know you already approved but will you please double check the latest couple commits and confirm whether this is what you had in mind?

**Review by zlav — APPROVED** [review]
Thanks for making those changes!

**Pull request merged**